ts/query.go

```diff
@@ -0,0 +1,411 @@
+		dsi.sampleIdx = 0
+	default:
+		dsi.valid = false
+	}
+	dsi.offset = data.offsetAt(dsi.sampleIdx)
+}
+
+// interpolatingIterator is used by dataSpan to iterate over all possible Sample
+// offsets. The iterator can provide sample values for any offset, interpolating
+// Sample values that do not exist in real data.
+type interpolatingIterator struct {
+	offset   int32            // Current offset within interpolated data set
+	nextReal dataSpanIterator // Next real sample with an offset >= iterator's offset
+	prevReal dataSpanIterator // Prev real sample with offset < iterator's offset
+}
+
+// advanceTo advances the iterator to the supplied offset.
+func (ii *interpolatingIterator) advanceTo(offset int32) {
+	ii.offset = offset
+	// Advance span iterators until nextReal has offset >= the interpolated
+	// offset.
+	for ii.nextReal.valid && ii.nextReal.offset < ii.offset {
+		ii.prevReal = ii.nextReal
+		ii.nextReal.advance()
+	}
+}
+
+// isValid returns true if this interpolatingIterator still points to valid data.
+func (ii *interpolatingIterator) isValid() bool {
+	return ii.nextReal.valid
+}
+
+// avg returns the average value at the current offset for this iterator.
+func (ii *interpolatingIterator) avg() float64 {
+	if !ii.isValid() {
+		return 0
+	}
+	if ii.nextReal.offset == ii.offset {
+		return ii.nextReal.sample().Average()
+	}
+	// Cannot interpolate if previous value is invalid.
+	if !ii.prevReal.valid {
+		return 0
+	}
+
+	// Linear interpolation of value at the current offset.
+	off := float64(ii.offset)
+	nextAvg := ii.nextReal.sample().Average()
+	nextOff := float64(ii.nextReal.offset)
+	prevAvg := ii.prevReal.sample().Average()
+	prevOff := float64(ii.prevReal.offset)
+	return prevAvg + (nextAvg-prevAvg)*(off-prevOff)/(nextOff-prevOff)
+}
+
+// newIterator returns an interpolating iterator for the given dataSpan. The
+// iterator is initialized to offset 0.
+func (ds *dataSpan) newIterator() interpolatingIterator {
+	if len(ds.datas) == 0 {
+		return interpolatingIterator{}
+	}
+
+	// The first data index necessarily contains the positive offset closest to 0. Use a
+	// binary search to find the lowest positive offset (or the zero offset).
+	data := ds.datas[0]
+	innerIdx := sort.Search(len(data.Samples), func(i int) bool {
+		return data.offsetAt(i) >= 0
+	})
+
+	iterator := interpolatingIterator{
+		offset: 0,
+		nextReal: dataSpanIterator{
+			dataSpan:  ds,
+			dataIdx:   0,
+			sampleIdx: innerIdx,
+			offset:    data.offsetAt(innerIdx),
+			valid:     true,
+		},
+	}
+
+	// If innerIdx > 0, then we can compute a "previous" iterator as well; this
+	// will let us interpolate a 0 value if it is not actually present in the
+	// data.
+	if innerIdx > 0 {
+		iterator.prevReal = dataSpanIterator{
+			dataSpan:  ds,
+			dataIdx:   0,
+			sampleIdx: innerIdx - 1,
+			offset:    data.offsetAt(innerIdx - 1),
+			valid:     true,
+		}
+	}
+
+	return iterator
+}
+
+// iteratorSet implements interesting methods over a set of
```

Comment:
I've significantly added to the comment, explaining why the iterators are advanced together and how the min heap faciliates that.